File: semester-1/programmieren/code/005/knights.py
# NEED TO IMPLEMENT THE WARNDORFF HEURISTIC
from time import time

board = []


def main():
    dimensions = 8
    start = input("enter starting coordinate: ")

    start_time = time()
    start = (int(start[1]) - 1, ord(start[0]) - 97)

    print(f"translated to program representation of board: {start}\n")
    for _ in range(0, int(dimensions)):
        board.append([0] * int(dimensions))

    knight_jump(start, 1)
    end_time = time()
    print_board(board)
    print(f"\ntime taken: {end_time - start_time} sec")


knight_jumps = {
    (2, 1),
    (1, 2),
    (2, -1),
    (1, -2),
    (-1, 2),
    (-1, -2),
    (-2, -1),
    (-2, 1),
}


def get_possible_jumps(current: tuple, bo):
    jumps = []
    for relative in knight_jumps:
        jump_to = (current[0] + relative[0], current[1] + relative[1])
        if not is_valid(jump_to, bo):
            continue
        degree = get_degree(jump_to, bo)

        pos = 0
        for i in range(len(jumps)):
            if get_degree(jumps[i], bo) < degree:
                pos += 1
        jumps.insert(pos, jump_to)

    possible_jumps = []
    for option in jumps:
        if not bo[option[0]][option[1]]:
            possible_jumps.append(option)

    return possible_jumps


def get_degree(coordinate, bo):
    count = 0
    for relative in knight_jumps:
        temp = (coordinate[0] + relative[0], coordinate[1] + relative[1])
        if is_valid(temp, bo):
            if not bo[temp[0]][temp[1]]:
                count += 1
    return count


def print_board(board):
    # reverse the board to have rows corresponding to actual chess board
    for row in reversed(board):
        row_format = "["
        for cell in row:
            row_format += "{0:02d}, ".format(cell)
        print(f"{row_format[:-2]}]")


def is_valid(coordinate, bo):
    return (coordinate[0] in range(0, len(bo[0]))) and (
        coordinate[1] in range(0, len(bo))
    )


def full(board):
    for row in board:
        for cell in row:
            if not cell:
                return False
    return True


def knight_jump(coordinates, move):
    board[coordinates[0]][coordinates[1]] = move

    if full(board):
        return True

    jumps = get_possible_jumps(coordinates, board)
    for jump in jumps:
        if knight_jump(jump, move + 1):
            return True

    board[coordinates[0]][coordinates[1]] = 0
    return False


if __name__ == "__main__":
    main()

File: semester-1/programmieren/code/005/knights.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>
#include <time.h>

#define DIMENSIONS 8

typedef struct
{
    int x;
    int y;
} tuple;

tuple KNIGHT_JUMPS[] = {
    {2, 1},
    {1, 2},
    {2, -1},
    {1, -2},
    {-1, 2},
    {-1, -2},
    {-2, -1},
    {-2, 1},
};

//-------------------------------------------------------------------

typedef struct
{
    tuple *data;       // pointer indicates where the memory of the struct instance lies
    uint64_t max_size; // the size of memory allocated
    uint64_t length;   // equivalent to java `array.length` attribute
} list_tuple;

// creates a new list filled with default values
list_tuple new_list(uint64_t size)
{
    list_tuple list = {
        malloc(size * sizeof(tuple)),
        size,
        0 // haha formatter, now you have to keep the closing brackets in the new line - I WON
    };

    return list;
}

void list_tuple_insert(list_tuple *list, int index, tuple element)
{
    // since we'll only use this to append knight jumps - 8 at maximum - we don't need to handle the case of length > max_size
    for (int i = list->length; i >= index; i--)
    {
        list->data[i] = list->data[i - 1];
    }
    list->data[index] = element;
    list->length++;
}


int BOARD[DIMENSIONS][DIMENSIONS];

//--------------------------------------------------------------------

void print_board()
{
    for (int i = 0; i < DIMENSIONS; i++)
    {
        for (int j = 0; j < DIMENSIONS; j++)
        {
            printf("%02d ", BOARD[i][j]);
        }
        printf("\n");
    }
}

bool is_valid(tuple coordinate)
{
    return (coordinate.x >= 0 && coordinate.x < DIMENSIONS) && (coordinate.y >= 0 && coordinate.y < DIMENSIONS);
}

bool is_full(int board[DIMENSIONS][DIMENSIONS])
{
    for (int i = 0; i < DIMENSIONS; i++)
    {
        for (int j = 0; j < DIMENSIONS; j++)
        {
            if (!board[i][j])
                return false;
        }
    }
    return true;
}

int get_degree(tuple coordinate)
{
    int count = 0;
    int length_knight_jumps = 8;
    for (int i = 0; i < length_knight_jumps; i++)
    {
        tuple temp = {
            coordinate.x + KNIGHT_JUMPS[i].x,
            coordinate.y + KNIGHT_JUMPS[i].y};
        if (is_valid(temp))
        {

            if (!BOARD[temp.x][temp.y])
            {
                count++;
            }
        }
    }

    return count;
}

list_tuple get_possible_jumps(tuple current)
{
    list_tuple jumps = new_list(8);
    int length_knight_jumps = 8;
    for (int i = 0; i < length_knight_jumps; i++)
    {
        tuple jump_to = {
            current.x + KNIGHT_JUMPS[i].x,
            current.y + KNIGHT_JUMPS[i].y};
        if (!is_valid(jump_to))
        {
            continue;
        }
        int degree = get_degree(jump_to);

        int pos = 0;
        for (int j = 0; j < jumps.length; j++)
        {
            if (get_degree(jumps.data[j]) < degree)
            {
                pos++;
            }
        }
        list_tuple_insert(&jumps, pos, jump_to);
    }

    // filters out squares that already have been visited
    list_tuple possible_jumps = new_list(8);
    for (int i = 0; i < jumps.length; i++)
    {
        if (!BOARD[jumps.data[i].x][jumps.data[i].y])
        {
            list_tuple_insert(&possible_jumps, possible_jumps.length, jumps.data[i]);
        }
    }
    free(jumps.data);
    return possible_jumps;
}

bool knight_jump(tuple coordinates, int move)
{
    BOARD[coordinates.x][coordinates.y] = move;
    if (is_full(BOARD))
        return true;

    list_tuple jumps = get_possible_jumps(coordinates);
    for (int i = 0; i < jumps.length; i++)
    {
        if (knight_jump(jumps.data[i], move + 1))
            return true;
    }
    BOARD[coordinates.x][coordinates.y] = 0;
    return false;
}

int main()
{
    char start_x, start_y;
    double time_start, time_end;
    printf("Enter starting coordinate (e.g. a1, g7): ");
    scanf("%c%c", &start_x, &start_y);
    // format coordinates to match array
    tuple start = {
        start_x - 'a',
        start_y - '1'};

    printf("%d, %d\n", start.x, start.y);
    time_start = (double)clock() / CLOCKS_PER_SEC;

    knight_jump(start, 1);

    time_end = (double)clock() / CLOCKS_PER_SEC;

    print_board(BOARD);
    printf("\n");
    printf("time taken: %lf\n", time_end - time_start);
    return 0;
}